native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCNameChecker.kt

```diff
@@ -0,0 +1,96 @@
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.jetbrains.kotlin.resolve.annotations.argumentValue
+import org.jetbrains.kotlin.resolve.checkers.DeclarationChecker
+import org.jetbrains.kotlin.resolve.checkers.DeclarationCheckerContext
+import org.jetbrains.kotlin.utils.addToStdlib.cast
+
+object NativeObjCNameChecker : DeclarationChecker {
+    private val objCNameFqName = FqName("kotlin.native.ObjCName")
+
+    override fun check(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        checkDeclaration(declaration, descriptor, context)
+        checkOverrides(declaration, descriptor, context)
+    }
+
+    private fun checkDeclaration(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        val objCNames = descriptor.getObjCNames().filterNotNull()
+        if (objCNames.isEmpty()) return
+        if (descriptor is CallableMemberDescriptor && descriptor.overriddenDescriptors.isNotEmpty()) {
+            objCNames.forEach {
+                val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(it.annotation) ?: declaration
+                context.trace.report(ErrorsNative.INAPPLICABLE_OBJC_NAME.on(reportLocation))
+            }
+        }
+        objCNames.forEach { checkObjCName(it, declaration, context) }
+    }
+
+    private fun checkObjCName(objCName: ObjCName, declaration: KtDeclaration, context: DeclarationCheckerContext) {
+        if (objCName.name == null && objCName.swiftName == null) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME.on(reportLocation))
+        }
+        val invalidNameChars = objCName.name?.toSet()?.intersect(NativeIdentifierChecker.invalidChars) ?: emptySet()
+        val invalidSwiftNameChars = objCName.swiftName?.toSet()?.intersect(NativeIdentifierChecker.invalidChars) ?: emptySet()
+        val invalidChars = invalidNameChars + invalidSwiftNameChars
+        if (invalidChars.isNotEmpty()) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_CHARACTERS_OBJC_NAME.on(reportLocation, invalidChars.joinToString("")))
+        }
+    }
+
+    private fun checkOverrides(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (descriptor !is CallableMemberDescriptor || descriptor.overriddenDescriptors.isEmpty()) return
+        val objCNames = descriptor.overriddenDescriptors.map { it.getFirstBaseDescriptor().getObjCNames() }
+        if (!objCNames.allNamesEquals()) {
+            context.trace.report(ErrorsNative.INCOMPATIBLE_OBJC_NAME_OVERRIDE.on(declaration))
+        }
```

Comment:
Hmm, I guess #4818 is also affected by this.

> There is a descriptor for Derived.foo though, so called "fake override".

At what point are those fake overrides available? Can we somehow get the fake overrides from the class declaration?

> `@Throws` checker partially handles the similar case of `@Throws` mismatch -- it would report `Derived2.foo`, not sure if this is actually enough (what if we didn't have `Derived2`?).

That doesn't sound ideal. In such cases I would probably expect an error on `Derived` similar to the "doesn't implement abstract function" error. How are those abstract function implementations checked (I couldn't find a check for it)? Such a check would have the same problem, right?

> I'm not proposing fixing this case in this PR, just starting a discussion. What do you think? Should we fix this eventually, and how?

Yeah IMO it should definitely be fixed (eventually). If it isn't fixed you example (without the `Derived2`) could result in unexpected names.
Ideally we could opt-in to check the fake overrides as well. But I am not sure if that is a (realistic) possibility.
I guess it all depends on if/where those fase overrides are available.

Reply:
> Hmm, I guess https://github.com/JetBrains/kotlin/pull/4818 is also affected by this.

Yes, probably.

> At what point are those fake overrides available? Can we somehow get the fake overrides from the class declaration?

Yes, they should be available among other members of `ClassDescriptor`.

> That doesn't sound ideal. In such cases I would probably expect an error on `Derived` 

Yes, I think so.

> How are those abstract function implementations checked (I couldn't find a check for it)? Such a check would have the same problem, right?

Maybe, but I don't see how this would be helpful here.

> Ideally we could opt-in to check the fake overrides as well. But I am not sure if that is a (realistic) possibility.
I guess it all depends on if/where those fase overrides are available.

I guess checking class declarations for such fake overrides in the same checker should do the trick.

Reply:
> Maybe, but I don't see how this would be helpful here.

Not sure, was hoping it would provide some inspiration 🙂.

> I guess checking class declarations for such fake overrides in the same checker should do the trick.

Cool, let me take a look at that.

Reply:
Alright this case is now also supported, the error will be reported on `Derived` in this case.
Let me know what you think of the implementation. If it's working as intended I will also add it to #4818.

Reply:
Thanks, the implementation looks neat! It might have a bug though: https://github.com/JetBrains/kotlin/pull/4815#discussion_r890017530